File: julia/NtToolBox/src/perform_haar_transf.jl
function perform_haar_transf(M, Jmin, dir)
    n = size(M)[1]
    Mf = copy(M)
    if dir == 1
        MW = copy(M)

        for j in Jmin : Int(log2(n))
            p = Int(n/2^(j-1))
            sel = 1 : p
            even = 1 : 2 : p
            odd = 2 : 2 : p
            # average/ difference along X
            MW[sel, sel, sel] = cat(1, (MW[even, sel, sel] + MW[odd, sel, sel])./sqrt(2), (MW[even, sel, sel] - MW[odd, sel, sel])./sqrt(2))
            # average/ difference along Y
            MW[sel, sel, sel] = cat(2, (MW[sel, even, sel] + MW[sel, odd, sel])./sqrt(2), (MW[sel, even, sel] - MW[sel, odd, sel])./sqrt(2))
            # average/ difference along Z
            MW[sel, sel, sel] = cat(3, (MW[sel, sel, even] + MW[sel, sel, odd])./sqrt(2), (MW[sel, sel, even] - MW[sel, sel, odd])./sqrt(2))
            Mf = copy(MW)
        end
    else

        M1 = copy(M)

        for j in Int(log2(n)) : -1 : Jmin
            p = Int(n/2^j)
            sel = 1 : p
            sel1 = 1:2*p
            selw = p+1:2*p
            even = 1:2:2*p
            odd = 2:2:2*p
            # average/ difference along X
            A = M1[sel, sel1, sel1]
            D = M1[selw, sel1, sel1]
            M1[even, sel1, sel1] = (A + D)./sqrt(2)
            M1[odd, sel1, sel1] = (A - D)./sqrt(2)
            # average/ difference along Y
            A = M1[sel1, sel, sel1]
            D = M1[sel1, selw, sel1]
            M1[sel1, even, sel1] = (A + D)./sqrt(2)
            M1[sel1, odd, sel1] = (A - D)./sqrt(2)
            # average/ difference along Z
            A = M1[sel1, sel1, sel]
            D = M1[sel1, sel1, selw]
            M1[sel1, sel1, even] = (A + D)./sqrt(2)
            M1[sel1, sel1, odd] = (A - D)./sqrt(2)
            Mf = copy(M1)
        end
    end

    return Mf
end

File: python/nt_toolbox/perform_haar_transf.py
import numpy as np

def perform_haar_transf(M, Jmin, dir):
    n = np.shape(M)[0]
    if dir == 1:
        MW = np.copy(M)
        
        for j in range(Jmin,int(np.log2(n))+1):
            p = int(n/2**(j-1))
            sel = np.arange(0,p)
            even = np.arange(0,p,2)
            odd = np.arange(1,p,2)
            # average/ difference along X
            MW[np.ix_(sel,sel,sel)] = np.concatenate(((MW[np.ix_(even,sel,sel)] + MW[np.ix_(odd,sel,sel)])/np.sqrt(2), (MW[np.ix_(even,sel,sel)] - MW[np.ix_(odd,sel,sel)])/np.sqrt(2)),0)
            # average/ difference along Y
            MW[np.ix_(sel,sel,sel)] = np.concatenate(((MW[np.ix_(sel,even,sel)] + MW[np.ix_(sel,odd,sel)])/np.sqrt(2), (MW[np.ix_(sel,even,sel)] - MW[np.ix_(sel,odd,sel)])/np.sqrt(2)),1)
            # average/ difference along Z
            MW[np.ix_(sel,sel,sel)] = np.concatenate(((MW[np.ix_(sel,sel,even)] + MW[np.ix_(sel,sel,odd)])/np.sqrt(2), (MW[np.ix_(sel,sel,even)] - MW[np.ix_(sel,sel,odd)])/np.sqrt(2)),2)
            Mf = MW
    else :
    
        M1 = np.copy(M)
        
        for j in range(int(np.log2(n)),Jmin-1,-1):
            p = int(n/2**j)
            sel = np.arange(0,p)
            sel1 = np.arange(0,2*p)
            selw = np.arange(p,2*p)
            even = np.arange(0,2*p,2)
            odd = np.arange(1,2*p,2)
            # average/ difference along X
            A = M1[np.ix_(sel,sel1,sel1)]
            D = M1[np.ix_(selw,sel1,sel1)]
            M1[np.ix_(even,sel1,sel1)] = (A + D)/np.sqrt(2)
            M1[np.ix_(odd,sel1,sel1)] = (A - D)/np.sqrt(2)
            # average/ difference along Y
            A = M1[np.ix_(sel1,sel,sel1)]
            D = M1[np.ix_(sel1,selw,sel1)]
            M1[np.ix_(sel1,even,sel1)] = (A + D)/np.sqrt(2)
            M1[np.ix_(sel1,odd,sel1)] = (A - D)/np.sqrt(2)
            # average/ difference along Z
            A = M1[np.ix_(sel1,sel1,sel)]
            D = M1[np.ix_(sel1,sel1,selw)]
            M1[np.ix_(sel1,sel1,even)] = (A + D)/np.sqrt(2)
            M1[np.ix_(sel1,sel1,odd)] = (A - D)/np.sqrt(2)
            Mf = M1

    return Mf